src/ext/VSExtension/ca/vsca.cpp

```diff
@@ -0,0 +1,360 @@
+    LPCWSTR pwzProperty;
+};
+
+static HRESULT InstanceIsGreater(
+    __in_opt ISetupInstance* pPreviousInstance,
+    __in DWORD64 qwPreviousVersion,
+    __in ISetupInstance* pCurrentInstance,
+    __in DWORD64 qwCurrentVersion
+    );
+
+static HRESULT ProcessInstance(
+    __in ISetupInstance* pInstance,
+    __in LPCWSTR wzProperty,
+    __in DWORD cComponents,
+    __in VS_COMPONENT_PROPERTY* rgComponents
+    );
+
+static HRESULT ProcessVS2017(
+    __in_opt ISetupInstance* pInstance,
+    __in DWORD64 qwVersion,
+    __in BOOL fComplete
+    );
+
+static HRESULT SetPropertyForComponent(
+    __in DWORD cComponents,
+    __in VS_COMPONENT_PROPERTY* rgComponents,
+    __in LPCWSTR wzComponent
+    );
+
+static VS_INSTANCE vrgInstances[] =
+{
+    { FILEMAKEVERSION(15, 0, 0, 0), FILEMAKEVERSION(15, 0xffff, 0xffff, 0xffff), ProcessVS2017 },
+};
+
+/******************************************************************
+ FindInstances - entry point for VS custom action to find instances
+
+*******************************************************************/
+extern "C" UINT __stdcall FindInstances(
+    __in MSIHANDLE hInstall
+    )
+{
+    HRESULT hr = S_OK;
+    UINT er = ERROR_SUCCESS;
+    BOOL fComInitialized = FALSE;
+    ISetupConfiguration* pConfiguration = NULL;
+    ISetupHelper* pHelper = NULL;
+    IEnumSetupInstances* pEnumInstances = NULL;
+    ISetupInstance* rgpInstances[1] = {};
+    ISetupInstance* pInstance = NULL;
+    ULONG cInstancesFetched = 0;
+    BSTR bstrVersion = NULL;
+    DWORD64 qwVersion = 0;
+
+    hr = WcaInitialize(hInstall, "VSFindInstances");
+    ExitOnFailure(hr, "Failed to initialize custom action.");
+
+    hr = ::CoInitialize(NULL);
+    ExitOnFailure(hr, "Failed to initialize COM.");
+
+    fComInitialized = TRUE;
+
+    hr = ::CoCreateInstance(__uuidof(SetupConfiguration), NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&pConfiguration));
+    if (REGDB_E_CLASSNOTREG != hr)
+    {
+        ExitOnFailure(hr, "Failed to initialize setup configuration class.");
+    }
+    else
+    {
+        WcaLog(LOGMSG_VERBOSE, "Setup configuration not registered; assuming no instances installed.");
+
+        hr = S_OK;
+        ExitFunction();
+    }
+
+    hr = pConfiguration->QueryInterface(IID_PPV_ARGS(&pHelper));
+    if (FAILED(hr))
+    {
+        WcaLog(LOGMSG_VERBOSE, "Setup configuration helpers not implemented; assuming Visual Studio 2017.");
+
+        qwVersion = FILEMAKEVERSION(15, 0, 0, 0);
+        hr = S_OK;
+    }
+
+    hr = pConfiguration->EnumInstances(&pEnumInstances);
+    ExitOnFailure(hr, "Failed to get instance enumerator.");
+
+    do
+    {
+        hr = pEnumInstances->Next(1, rgpInstances, &cInstancesFetched);
+        if (SUCCEEDED(hr) && cInstancesFetched)
+        {
+            pInstance = rgpInstances[0];
+            if (pInstance)
+            {
+                if (pHelper)
+                {
+                    hr = pInstance->GetInstallationVersion(&bstrVersion);
```

Comment:
Looks like this could overwrite bstrVersion from a previous iteration of the loop, leaking the prior string.